lib/protocol/protocol.go

```diff
@@ -801,22 +809,42 @@ func (c *rawConnection) shouldCompressMessage(msg message) bool {
 	}
 }
 
+// Close is called when the connection is regularely closed and thus the Close
+// BEP message is sent before terminating the actual connection. The error
+// argument specifies the reason for closing the connection.
+func (c *rawConnection) Close(err error) {
+	c.once.Do(func() {
+		done := make(chan struct{})
+		c.sendClose <- asyncMessage{&Close{err.Error()}, done}
+		<-done
+
+		go c.commonClose(err)
```

Comment:
Some sort of comment on why we `go` this? I have a vague feeling it's because this might be called from a protocol callback and there are potential deadlock issues?

Reply:
Clear like this?

Reply:
Why wouldn't we want an internal close here? Is it valid to close the connection multiple times? I assume that would panic on closing c.closed a second time. At which point the sync.once can move into the commonClose and commonClose can become internalClose?

Reply:
You lost me there. However you made me look at it again and I found a way this can deadlock and solving that indeed gets rid of `commonClose` (but introduces a second once). PR incoming